src/transform/src/dataflow.rs

```diff
@@ -1132,25 +1144,72 @@ impl IndexUsageContext {
         old_contexts: &Vec<IndexUsageContext>,
         keys_to_add: &Vec<Vec<MirScalarExpr>>,
     ) -> Vec<IndexUsageContext> {
-        let mut contexts = old_contexts.clone();
-        if contexts.is_empty() {
+        let old_contexts = if old_contexts.is_empty() {
             // No join above us, and we are not at the root. Why does this ArrangeBy even exist?
             soft_panic_or_log!("CollectIndexRequests encountered a dangling ArrangeBy");
             // Anyhow, let's create a context with a `DanglingArrangeBy` index usage, so that we
             // have a place to note down the requested keys below.
-            contexts = IndexUsageContext::from_usage_type(IndexUsageType::DanglingArrangeBy);
-        }
-        for context in contexts.iter_mut() {
-            if context.requested_keys.is_none() {
-                context.requested_keys = Some(BTreeSet::new());
-            }
-            context
-                .requested_keys
-                .as_mut()
-                .unwrap()
-                .extend(keys_to_add.iter().cloned());
-        }
-        contexts
+            IndexUsageContext::from_usage_type(IndexUsageType::DanglingArrangeBy)
+        } else {
+            old_contexts.clone()
+        };
+        old_contexts
+            .into_iter()
+            .flat_map(|old_context| {
+                if !matches!(
+                    old_context.usage_type,
+                    IndexUsageType::DeltaJoin(DeltaJoinIndexUsageType::Unknown)
+                ) {
+                    // If it's not an unknown delta join usage, then we simply note down the new
+                    // keys into `requested_keys`.
+                    let mut context = old_context.clone();
+                    if context.requested_keys.is_none() {
+                        context.requested_keys = Some(BTreeSet::new());
+                    }
+                    context
+                        .requested_keys
+                        .as_mut()
+                        .unwrap()
+                        .extend(keys_to_add.iter().cloned());
+                    vec![context]
```

Comment:
You can avoid the allocation here and below!
```diff
diff --git a/src/transform/src/dataflow.rs b/src/transform/src/dataflow.rs
index 25eefd7959..9a3dfbe711 100644
--- a/src/transform/src/dataflow.rs
+++ b/src/transform/src/dataflow.rs
@@ -1171,7 +1171,7 @@ impl IndexUsageContext {
                         .as_mut()
                         .unwrap()
                         .extend(keys_to_add.iter().cloned());
-                    vec![context]
+                    Some(context).into_iter().chain(None)
                 } else {
                     // If it's an unknown delta join usage, then we need to figure out whether this
                     // is a full scan or a lookup.
@@ -1199,13 +1199,15 @@ impl IndexUsageContext {
                         .cloned()
                         .collect_vec();
                     if lookup_keys.is_empty() {
-                        vec![full_scan_context]
+                        Some(full_scan_context).into_iter().chain(None)
                     } else {
                         let lookup_context = IndexUsageContext {
                             requested_keys: Some(lookup_keys.into_iter().collect()),
                             usage_type: IndexUsageType::DeltaJoin(DeltaJoinIndexUsageType::Lookup),
                         };
-                        vec![full_scan_context, lookup_context]
+                        Some(full_scan_context)
+                            .into_iter()
+                            .chain(Some(lookup_context))
                     }
                 }
             })
```

Reply:
Thanks!